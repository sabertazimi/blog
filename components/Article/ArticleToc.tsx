'use client'

import { useEffect, useState } from 'react'
import styles from './ArticleToc.module.css'
import Anchor from '@/components/Anchor'

interface Props {
  slug: string
  offset?: number
}

interface TocItem {
  key: string | number
  href: string
  title: string
}

function ArticleToc({ slug, offset = 150 }: Props): JSX.Element {
  const [tocItems, setTocItems] = useState<TocItem[]>([])

  useEffect(() => {
    const items = document.querySelectorAll('h2.ant-typography')
    const tocItems = Array.from(items).map(item => ({
      key: `#${item.id}`,
      href: `#${item.id}`,
      title: item.textContent,
    })) as TocItem[]
    setTocItems(tocItems)
  }, [slug])

  return (
    <div className={styles.toc}>
      <Anchor offsetTop={offset} items={tocItems} />
    </div>
  )
}

export default ArticleToc
