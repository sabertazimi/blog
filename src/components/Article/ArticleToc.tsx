import { CloseOutlined, MenuFoldOutlined } from '@ant-design/icons';
import { Button, Drawer } from 'antd';
import classNames from 'classnames';
import React, { useCallback, useState } from 'react';
import * as styles from './ArticleToc.module.css';

interface Props {
  toc: string;
}

const ArticleToc = ({ toc }: Props): JSX.Element => {
  const [tocVisible, setTocVisible] = useState(false);

  const handleClick = useCallback(
    () => setTocVisible(tocVisible => !tocVisible),
    []
  );

  return (
    <div
      className={classNames(
        'fixed hidden m-0 bg-transparent',
        'top-8 right-28 z-max',
        'md:block'
      )}
      role="navigation"
    >
      <Button
        className="button-primary"
        shape="circle"
        icon={tocVisible ? <CloseOutlined /> : <MenuFoldOutlined />}
        size="large"
        onClick={handleClick}
      />
      <Drawer
        className="font-extrabold text-dark"
        title="Table of Contents"
        placement="right"
        closable={false}
        onClose={handleClick}
        visible={tocVisible}
      >
        <div
          className={classNames('transition transform-gpu', styles.toc, {
            'scale-100': tocVisible,
            'scale-0': !tocVisible,
          })}
          dangerouslySetInnerHTML={{ __html: toc }}
          onClick={handleClick}
        />
      </Drawer>
    </div>
  );
};

export default ArticleToc;
