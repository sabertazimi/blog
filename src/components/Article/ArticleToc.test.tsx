import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { create } from 'react-test-renderer';
import ArticleToc from './ArticleToc';

const toc = 'Post Table of Contents';

describe('ArticleToc', () => {
  test('should render correctly (snapshot)', () => {
    const tree = create(<ArticleToc toc={toc} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should expand ToC when clicked', () => {
    render(<ArticleToc toc={toc} />);

    fireEvent.click(screen.getByRole('button'));

    expect(screen.getByText(toc)).toBeInTheDocument();
  });
});
