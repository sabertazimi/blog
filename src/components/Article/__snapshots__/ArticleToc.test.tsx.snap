// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleToc should render correctly (snapshot) 1`] = `
<div
  className="fixed hidden m-0 bg-transparent top-8 right-28 z-max md:block"
  role="navigation"
>
  <button
    className="ant-btn ant-btn-circle ant-btn-default ant-btn-lg ant-btn-icon-only button-primary"
    onClick={[Function]}
    type="button"
  >
    <span
      aria-label="menu-fold"
      className="anticon anticon-menu-fold"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="menu-fold"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 000 13.8z"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`ArticleToc should render correctly (snapshot) 2`] = `
<div
  className="fixed hidden m-0 bg-transparent top-8 right-28 z-max md:block"
  role="navigation"
>
  <button
    className="ant-btn ant-btn-circle ant-btn-default ant-btn-lg ant-btn-icon-only button-primary"
    onClick={[Function]}
    type="button"
  >
    <span
      aria-label="close"
      className="anticon anticon-close"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="close"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
        />
      </svg>
    </span>
  </button>
  <div>
    <div
      id="content"
    >
      <div
        className="ant-drawer ant-drawer-right font-extrabold text-dark"
        onTransitionEnd={[Function]}
        style={
          Object {
            "transform": undefined,
            "zIndex": undefined,
          }
        }
        tabIndex={-1}
      >
        <div
          className="ant-drawer-mask"
          onClick={[Function]}
          style={Object {}}
        />
        <div
          className="ant-drawer-content-wrapper"
          style={
            Object {
              "height": undefined,
              "msTransform": "translateX(100%)",
              "transform": "translateX(100%)",
              "width": "378px",
            }
          }
        >
          <div
            className="ant-drawer-content"
          >
            <div
              className="ant-drawer-wrapper-body"
              onTransitionEnd={[Function]}
              style={Object {}}
            >
              <div
                className="ant-drawer-header"
              >
                <div
                  className="ant-drawer-header-title"
                >
                  <div
                    className="ant-drawer-title"
                  >
                    Table of Contents
                  </div>
                </div>
              </div>
              <div
                className="ant-drawer-body"
              >
                <div
                  className="transition transform-gpu scale-100"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "Post Table of Contents",
                    }
                  }
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      data-target-tag-name="DIV"
      id="target"
    />
  </div>
</div>
`;
